Edit tools shift

yourealwaysbe opened the pull request:
Another version of #48, #50, and #52.  This time shift has to be pressed to enable the split/join behaviour.  Addressing @rnorris's concerns in #50.

It could be improved by changing the cursor icon when shift is held.  I could not see a nice easy way of doing that.  If there's a standard way, let me know!

The route finder tools also has the shift-split/join behaviour.  This has the disadvantage that if a user does a join, but the returned route is not what they hoped, they cannot directly undo.  They can, however, use the split/join behaviour to replace the planned section.

The ctrl-click behaviour is restored here (i clobbered it in #52, but that can be fixed).

rnorris commented:
This is my preferred way to progress.

In my quick testing (with routes), I was able to split/join a route - but the route ended up with more than one segment - which shouldn't be allowed.
I think somewhere this needs to be added:

    // Routes can only have one segment
    if ( trk->is_route ) {
      (void)vik_track_merge_segments ( trk );
    }

NB This may need to be in all your solutions (should we decide to pick one of the others)
NB2 Also see previous commit SHA: ae3e3c81665358ec1b46b0adde8ce040818eaab5 about this.

yourealwaysbe commented:
I prefer this option also.

I was not able to create multiple segments with split/join.  Which tool did you use?  The route-finder tool has the issue of creating multiple segments.  I can fix it using your code.  However, the first trackpoint, which is duplicated, still amounts to two new segments.  Is this intended?

yourealwaysbe commented:
I implemented changing the cursor to s crosshair when shift is pressed.  This required adding a key_release function to the tool interface.  Also, i'm storing the crosshair_cursor in the viktrwlayer.c.  Maybe this would be better in vikwindow.c like vik_window_set_busy_cursor (vik_window_set_crosshair_cursor?).

What should the newsegment behaviour be for tracks?  If the tool is used to append one track to another, should there be two segments?  At the moment the appended track doesn't get a newsegment (i added code to explicitly prevent this).

rnorris commented on src/viktrwlayer.c:
```diff
@@ -10274,10 +10285,24 @@ static gboolean tool_edit_track_key_press ( VikTrwLayer *vtl, GdkEventKey *event
+    vik_window_clear_busy_cursor ( vw );
+  }
+}
+
```
This tool_edit_track_key_release () should be returning a value

yourealwaysbe replied:
Sorry for that.  I've fixed it and updated the pull request.  Thanks.

yourealwaysbe commented:
I've updated the merge request to fix the missing return issue.  There is a conflict now since master has changed in the meantime.  I have resolved it, but let me know if another way of fixing the conflict is preferred.

rnorris commented:
-      <toolitem action='CreateRoute'/>"
    +      <toolitem action='EditRoute'/>"

    -    { "CreateRoute", "vik-icon-Create Route", N_("Create _Route"), "<control><shift>B", N_("Create Route"), 0 },
    +    { "EditRoute", "vik-icon-Edit Route", N_("Edit _Route"), "<control><shift>B", N_("Edit Route"), 0 },

This comment focuses on Routes, but of course applies to Track functionality as well.


I would like to consider as much backwards compatibility as possible, so unfortunately the simple code change here of changing CreateRoute -> EditRoute as used in the toolitem action name means if someone has an existing customized toolbar setup would lose the buttons for 'CreateRoute' since the action no longer exists.

So some options are:

1. Tough luck for the end user. You have to manually add in the new actions to your customized toolbar.
A note can be made in the Manual / Release notes in case any one cared to read them.
Obviously easier for the developers as we don't have to do any more work.
Possibly the user would lose a keyboard shortcut if one was setup too.

2. Keep the tool action names as before e.g. 'CreateRoute' but all the labels and functions change to the 'edit' version to reflect
the new capability. So it's only the internal text name that is not fully describing it's functionality but keeps backwards capability.
No change for end users, not much effort for developer(s) other than putting in suitable comments about why this name hasn't changed since it preserves the backwards compatibility.
This would probably preserve a keyboard shortcut if one was setup too.

3. Keep the old tools *and* add the new tools.
This keeps the backwards compatibility, but doesn't necessary offer the new tools to users.
For developers we would have more and/or duplicated code.
Possiby could then state the old tool is depreciated and then delete it in a future subsequent release (e.g. v1.8)


ATM I think approach 2. is the best option.

yourealwaysbe commented:
I agree about option 2.  I've pushed a new version.

I keep forgetting the user config toolbar thing!

rnorris commented:
NB For the _Cutter_ tool, my previous method is to use the Select Tool to select a trackpoint or routepoint and then on the right click menu use Split->Split at Trackpoint.

Would it be better to call it a __Splitter__ tool, to mirror the Split terminology used in the menu option (and also the  'Split at Marker' on the Properties Dialog graphs)?
It might also reduce potential confusion for the difference between __Cutter__ and the normal edit __Cut__

Probably would then use Shift+Ctrl+L as that I think is the only unused related key combination.

yourealwaysbe commented:
Cool -- i've renamed to splitter.  I would have no objections if you decide a dedicated splitter tool is too much clutter.
